src/core/lib/iomgr/exec_ctx.h

```diff
@@ -227,6 +230,48 @@ class ExecCtx {
   ExecCtx* last_exec_ctx_ = Get();
 };
 
+/** Application-callback execution context.
+ *  A bag of data that collects information along a callstack.
+ *  It is created on the stack at core entry points, and stored internally
+ *  as a thread-local variable.
+ *
+ *  There are three key differences between this structure and ExecCtx:
+ *    1. ApplicationCallbackExecCtx builds a list of application-level
+ *       callbacks, but ExecCtx builds a list of internal callbacks to invoke.
+ *    2. If more than one ApplicationCallbackExecCtx is created on the thread's
```

Comment:
If we're adding a new type of ExecCtx, can we start from the beginning by requiring that only one instance is present in a given thread at a given time?  That way, we won't have to clean this up later the way we're going to have to do with the normal ExecCtx.

Reply:
This is essentially impossible since ApplicationCallbackExecCtx causes application code to run, which can re-enter core which would then need another ApplicationCallbackExecCtx. In practice because nothing ever gets enqueued on the next one, the maximum depth of this will only ever be 2.

Reply:
I've added clarifications here (and re-ordered the bullet items)